Samples/Desktop/D3D12HelloWorld/src/HelloVADecode/D3D12HelloVADecode.cpp

```diff
@@ -0,0 +1,634 @@
+    // sample illustrates how to use fences for efficient resource usage and to
+    // maximize GPU utilization.
+
+    // Signal and increment the fence value.
+    const UINT64 fence = m_fenceValue;
+    ThrowIfFailed(m_commandQueue->Signal(m_fence.Get(), fence));
+    m_fenceValue++;
+
+    // Wait until the previous frame is finished.
+    if (m_fence->GetCompletedValue() < fence)
+    {
+        ThrowIfFailed(m_fence->SetEventOnCompletion(fence, m_fenceEvent));
+        WaitForSingleObject(m_fenceEvent, INFINITE);
+    }
+
+    m_frameIndex = m_swapChain->GetCurrentBackBufferIndex();
+}
+
+
+void D3D12HelloVADecode::InitVADecContext()
+{
+    VAStatus va_status = vaCreateConfig(
+        m_vaDisplay,
+        VAProfileH264Main,
+        VAEntrypointVLD,
+        nullptr,
+        0,
+        &m_VADecConfigId);
+    ThrowIfFailed(va_status, "vaCreateConfig");
+
+    va_status = vaCreateContext(
+        m_vaDisplay,
+        m_VADecConfigId,
+        GetWidth(),
+        GetHeight(),
+        VA_PROGRESSIVE,
+        m_VARenderTargets,
+        FrameCount,
+        &m_VADecContextId);
+    ThrowIfFailed(va_status, "vaCreateContext");
+
+    va_status = vaCreateBuffer(
+        m_vaDisplay,
+        m_VADecContextId,
+        VAPictureParameterBufferType,
+        sizeof(VAPictureParameterBufferH264),
+        1,
+        (void*) g_PicParams_AVC,
+        &m_VADecPipelineBufferId[VA_H264DEC_BUFFER_INDEX_PIC]);
+    ThrowIfFailed(va_status, "vaCreateBuffer");
+
+    va_status = vaCreateBuffer(
+        m_vaDisplay,
+        m_VADecContextId,
+        VAIQMatrixBufferType,
+        sizeof(VAIQMatrixBufferH264),
+        1,
+        (void*) g_Qmatrix_AVC,
+        &m_VADecPipelineBufferId[VA_H264DEC_BUFFER_INDEX_QMATRIX]);
+    ThrowIfFailed(va_status, "vaCreateBuffer");
+
+    va_status = vaCreateBuffer(
+        m_vaDisplay,
+        m_VADecContextId,
+        VASliceDataBufferType,
+        sizeof(g_Bitstream_AVC),
+        1,
+        (void*)g_Bitstream_AVC,
+        &m_VADecPipelineBufferId[VA_H264DEC_BUFFER_INDEX_COMPRESSED_BIT]);
+    ThrowIfFailed(va_status, "vaCreateBuffer");
+
+    va_status = vaCreateBuffer(
+        m_vaDisplay,
+        m_VADecContextId,
+        VASliceParameterBufferType,
+        sizeof(VASliceParameterBufferH264),
+        1,
+        (void*) g_SlcParams_AVC,
+        &m_VADecPipelineBufferId[VA_H264DEC_BUFFER_INDEX_SLICE]);
+    ThrowIfFailed(va_status, "vaCreateBuffer");
+}
+
+void D3D12HelloVADecode::PerformVADecodeFrame(VASurfaceID dst_surface)
+{
+    VAStatus va_status;
+    va_status = vaBeginPicture(m_vaDisplay, m_VADecContextId, dst_surface);
+    ThrowIfFailed(va_status, "vaBeginPicture");
+    
+    vaRenderPicture(m_vaDisplay, m_VADecContextId, m_VADecPipelineBufferId, _countof(m_VADecPipelineBufferId));
+
+    va_status = vaEndPicture(m_vaDisplay, m_VADecContextId);
+    ThrowIfFailed(va_status, "vaEndPicture");
+
+    // Flush work on GPU and finish for completion on GPU
```

Comment:
Does this basically mean 'wait for completion on GPU'?

Reply:
yes, I'll reword the comment for more clarity :)